internal/pkg/skuba/addons/addons.go

```diff
@@ -0,0 +1,265 @@
+	if err != nil {
+		return "", errors.Wrap(err, "could not parse manifest template")
+	}
+	var rendered bytes.Buffer
+	err = template.Execute(&rendered, addonRenderContext{
+		Addon:  addon,
+		Config: addonConfiguration,
+	})
+	if err != nil {
+		return "", errors.Wrapf(err, "could not render manifest template")
+	}
+	return rendered.String(), nil
+}
+
+func (addon Addon) IsPresentForClusterVersion(clusterVersion *version.Version) bool {
+	return kubernetes.AddonVersionForClusterVersion(addon.addon, clusterVersion) != nil
+}
+
+func (addon Addon) HasToBeApplied(addonConfiguration AddonConfiguration, skubaConfiguration *skuba.SkubaConfiguration) (bool, error) {
+	if !addon.IsPresentForClusterVersion(addonConfiguration.ClusterVersion) {
+		// TODO (ereslibre): this logic can be triggered if some registered
+		//       addons are not supported in all Kubernetes versions. Either:
+		//
+		//   a) When rendering all addons on `skuba cluster init`, we skip those
+		//      that don't apply to the chosen Kubernetes version.
+		//
+		//   b) When running `skuba addon upgrade apply`; in this case (hence the
+		//      TODO), should we trigger a deletion of the addons that are not present
+		//      in the new version but were present on the old Kubernetes version? For
+		//      now, just return that it doesn't have to be applied.
+		return false, nil
+	}
+	if skubaConfiguration.AddonsVersion == nil {
+		return true, nil
+	}
+	currentAddonVersion, found := skubaConfiguration.AddonsVersion[addon.addon]
+	if !found {
+		return true, nil
+	}
+	addonVersion := kubernetes.AddonVersionForClusterVersion(addon.addon, addonConfiguration.ClusterVersion)
+	return !addonVersionLower(addonVersion, currentAddonVersion), nil
+}
+
+func (addon Addon) needsRender(applyBehavior ApplyBehavior) bool {
+	if applyBehavior == AlwaysRender {
+		return true
+	}
+	_, err := os.Stat(addon.manifestPath())
+	return err != nil
+}
+
+func (addon Addon) addonPath() string {
+	return filepath.Join(skubaconstants.AddonsDir(), string(addon.addon))
+}
+
+func (addon Addon) manifestPath() string {
+	return filepath.Join(addon.addonPath(), fmt.Sprintf("%s.yaml", addon.addon))
+}
+
+func (addon Addon) Write(addonConfiguration AddonConfiguration) error {
+	addonManifest, err := addon.Render(addonConfiguration)
+	if err != nil {
+		return errors.Wrapf(err, "unable to render %s addon template", addon.addon)
+	}
+	os.MkdirAll(addon.addonPath(), 0700)
+	if err := ioutil.WriteFile(addon.manifestPath(), []byte(addonManifest), 0600); err != nil {
+		return errors.Wrapf(err, "unable to write %s addon rendered template", addon.addon)
+	}
+	return nil
+}
+
+func (addon Addon) Apply(addonConfiguration AddonConfiguration, skubaConfiguration *skuba.SkubaConfiguration, applyBehavior ApplyBehavior) error {
+	klog.V(1).Infof("applying %q addon", addon.addon)
+	if addon.callbacks != nil {
+		if err := addon.callbacks.BeforeApply(addonConfiguration, skubaConfiguration); err != nil {
```

Comment:
If I understand this correctly, if one addon fails but another doesn't, then we will get anyway an error but at least the successful addons won't have to be re-applied. Is that correct? Or this is an "all or nothing" situation? I'm a bit lost here :sweat_smile:

Reply:
Exactly, if we get an error applying one addon, we will still continue trying to apply the rest, `DeployAddons` has the loop, we will `Errorf` but continue with the loop.